### PR TITLE
feat: JetStream auto-reconnection

### PR DESCRIPTION
**Pull request by whynowy**
This PR implements our own reliable JetStream auto-reconnection mechanism.

Closes: https://github.com/numaproj/numaflow/issues/52

### CODE REVIEW

File: pkg/shared/clients/jetstream/nats_conn.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	"time"
+
+	"github.com/nats-io/nats.go"
+)
+
+// NatsConn is a wrapper of nats.Conn, which implements our own magic for auto reconnection.
+type NatsConn struct {
+	Conn *nats.Conn
+
+	// pintConext is a dedicated JetStreamContext used to check if connection is OK.
```
Comment (vigith) [inline pkg/shared/clients/jetstream/nats_conn.go:15]:
```suggestion
	// pingConext is a dedicated JetStreamContext used to check if connection is OK.
```

File: pkg/shared/clients/jetstream/in_cluster_jetstream_client.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	for _, o := range opts {
+		if o != nil {
+			o(options)
+		}
+	}
+	nc, err := isc.connect(ctx)
+	if err != nil {
+		return nil, err
+	}
+	result := NewNatsConn(nc)
```
Comment (vigith) [inline pkg/shared/clients/jetstream/in_cluster_jetstream_client.go:64]: rename `result` to `custom...`

File: pkg/shared/clients/jetstream/options.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+func defaultJetStreamClientOptions() *jsClientOptions {
+	return &jsClientOptions{
+		reconnect:               true,
+		connectionCheckInterval: 6 * time.Second,
+	}
+}
+
+type JetStreamClientOption func(*jsClientOptions)
+
+// Set no auto reconnect
```
Comment (vigith) [inline pkg/shared/clients/jetstream/options.go:24]:
```suggestion
// NoReconnect sets to no auto reconnect
```

Comment (vigith) [reply pkg/shared/clients/jetstream/options.go:24]: couple of other places too

File: pkg/isb/jetstream/reader.go [hunk 4]
```diff
@@ -32,9 +33,54 @@ type jetStreamReader struct {
...
+
+	connectAndSubscribe := func() (*jsclient.NatsConn, *jsclient.JetStreamContext, *nats.Subscription, error) {
+		conn, err := client.Connect(ctx, jsclient.ReconnectHandler(func(c *jsclient.NatsConn) {
+			if result.js == nil {
+				log.Error("JetStreamContext is nil")
+				return
+			}
+			var e error
+			_ = wait.ExponentialBackoffWithContext(ctx, wait.Backoff{
+				Steps:    5,
```
[the remark this answers is not in the document]
Comment (whynowy) [reply pkg/isb/jetstream/reader.go:63]: It's good enough I think.

File: pkg/isb/jetstream/reader.go [hunk 5]
```diff
@@ -32,9 +33,54 @@ type jetStreamReader struct {
...
+					log.Info("Re-subscribed to the stream successfully")
+					return true, nil
+				}
+			})
+			if e != nil {
+				// Let it panic to start over
+				log.Fatalw("Failed to re-subscribe after retries", zap.Error(e))
+			}
+		}), jsclient.DisconnectErrHandler(func(nc *jsclient.NatsConn, err error) {
+			log.Error("Nats JetStream connection lost")
```
[the remark this answers is not in the document]
Comment (whynowy) [reply pkg/isb/jetstream/reader.go]: done.

File: pkg/shared/clients/jetstream/nats_conn.go (hunk 1 above)
Comment (whynowy) [reply pkg/shared/clients/jetstream/nats_conn.go]: done.

File: pkg/shared/clients/jetstream/in_cluster_jetstream_client.go (hunk 2 above)
Comment (whynowy) [reply pkg/shared/clients/jetstream/in_cluster_jetstream_client.go]: done.

File: pkg/shared/clients/jetstream/options.go (hunk 3 above)
Comment (whynowy) [reply pkg/shared/clients/jetstream/options.go]: done.

**Review by vigith — APPROVED** [review]

**Pull request merged**